### PR TITLE
Add safety checking infrastructure for text generation

### LINKED ISSUE #41740: Safety Checking Infrastructure for Text Generation
### Feature request

Add moderation to text generation models, similar to the approaches used in production LLMs (e.g. ChatGPT, Gemini, Claude) so that open-source models can have built-in content filtering.

This could provide
- Abstract base class for implementing safety checkers/classifiers
- Integration hooks during generation (logits processing and stopping criteria)
- Pipeline support for safe generation
- Configuration for safety settings

### Motivation

While production LLMs have built-in safety moderation systems, they are often insufficient and can lead to unexpected harmful behavior, especially over long conversations. As open-source text generation models become more capable and widely used, mitigating harm and ensuring user safety is a feature that should be built in. As far as I am aware, there is currently no built-in infrastructure to support this. The most effective approaches involve moderation during inference, which is a non-trivial feature for Transformers users to implement on their own. In addition, allowing for the configuration of safety with custom settings and classifiers can allow users to avoid harm in more specialized contexts than commercial LLMs currently address.

### Your contribution

I would like to work on a PR for this feature if there is interest. It seems that Diffusers' `StableDiffusionSafetyChecker` provides precedent, and `LogitsProcessor` and `StoppingCriteria` would support the development of this feature.

### PR DESCRIPTION
**Pull request by rice-e**
**Draft PR:** Core implementation complete. Seeking feedback on design and approach before finalizing. Thank you!

# What does this PR do?

Adds safety checking infrastructure for text generation. Provides base classes, configuration, and processors that integrate with the generation pipeline. Users implement their own safety checkers for specific needs (harm, bias, PII, etc.).

Fixes #41740

# Motivation
As stated in the issue I opened, while production LLMs have built-in safety moderation systems, they are often insufficient and can lead to unexpected harmful behavior, especially over long conversations. As open-source text generation models become more capable and widely used, mitigating harm and ensuring user safety is a feature that should be built in. As far as I am aware, there is currently no built-in infrastructure to support this. The most effective approaches involve moderation during inference, which is a non-trivial feature for Transformers users to implement on their own. In addition, allowing for the configuration of safety with custom settings and classifiers can allow users to avoid harm in more specialized contexts than commercial LLMs currently address.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. Currently being discussed in #41740
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

### CODE REVIEW

**Comment by qgallouedec** [comment]
see https://github.com/huggingface/transformers/issues/41740#issuecomment-5060062462
